rama-net/src/client/pool.rs

```diff
@@ -0,0 +1,618 @@
+                _slot: active_slot,
+                pooled_conn,
+                pool,
+            };
+            return Ok(GetConnectionOrCreate::LeasedConnection(leased_conn));
+        };
+
+        // If we have an active slot we should always be able to get a pool slot. Unless our pool
+        // is totally full, and in that case we just need to drop one connection and use that slot instead
+        let pool_slot = match self.inner.total_slots.clone().try_acquire_owned() {
+            Ok(pool_permit) => PoolSlot(pool_permit),
+            Err(_) => {
+                let pooled_conn = self.inner.connections.lock().pop_back().context(
+                    "connections vec cannot be empty if total slots doesn't have permit available",
+                )?;
+                pooled_conn.slot
+            }
+        };
+
+        Ok(GetConnectionOrCreate::AddConnection(move |conn: C| {
+            LeasedConnection {
+                _slot: active_slot,
+                pool,
+                pooled_conn: Some(PooledConnection {
+                    conn,
+                    id: id.clone(),
+                    slot: pool_slot,
+                }),
+            }
+        }))
+    }
+}
+
+/// [`ReqToConnId`] is used to convert a `Request` to a connection ID. These IDs are
+/// not unique and multiple connections can have the same ID. IDs are used to filter
+/// which connections can be used for a specific Request in a way that is indepent of
+/// what a Request is.
+pub trait ReqToConnId<Request>: Sized + Send + Sync + 'static
+where
+    Request: Send + 'static,
+{
+    type ConnId: Send + Sync + PartialEq + Clone + 'static;
+
+    fn id(&self, request: Request) -> (Request, Self::ConnId);
+}
+
+impl<Request, ConnId, F> ReqToConnId<Request> for F
+where
+    F: Fn(Request) -> (Request, ConnId) + Send + Sync + 'static,
+    Request: Send + 'static,
+    ConnId: Send + Sync + PartialEq + Clone + 'static,
+{
+    type ConnId = ConnId;
+
+    fn id(&self, request: Request) -> (Request, Self::ConnId) {
+        self(request)
+    }
+}
+
+/// [`PooledConnector`] is a connector that will keep connections around in a local pool
+/// so they can be reused later. If no connections are available for a specifc `id`
+/// it will create a new one.
+pub struct PooledConnector<S, C, Request: Send + 'static, R: ReqToConnId<Request>> {
+    inner: S,
+    pool: Pool<C, R::ConnId>,
+    req_to_conn_id: R,
+    wait_for_pool_timeout: Option<Duration>,
+}
+
+impl<S, C, Request: Send + 'static, R: ReqToConnId<Request>> PooledConnector<S, C, Request, R> {
+    pub fn new(
+        inner: S,
+        pool: Pool<C, R::ConnId>,
+        req_to_conn_id: R,
+    ) -> PooledConnector<S, C, Request, R> {
+        PooledConnector {
+            inner,
+            pool,
+            req_to_conn_id,
+            wait_for_pool_timeout: None,
+        }
+    }
+
+    pub fn with_wait_for_pool_timeout(mut self, timeout: Option<Duration>) -> Self {
```

Comment:
I tend to call these kind of things `maybe_with_x` and also provide a version `with_x` without the Option wrpaper, and even a `set_x` version for `&mut self` (without the option wrapper).